### ISSUE TITLE
MPTCP support

### ISSUE THREAD

**Issue by voidpointertonull** [syncthing/syncthing#8846]
Initial multipath TCP support was merged into golang recently which should make its usage quite convenient:
https://github.com/golang/go/issues/56539

Given that MPTCP is essentially an extension of TCP, basic support should only need opting into the extension at the time of socket creation. The kernel definitely supports graceful fallback to TCP in case the other side doesn't signal MPTCP support or a bogus middlebox breaks support, and apparently the golang implementation also adds fallback, even though the test case also explicitly checks for success:
https://github.com/golang/go/commit/383a4e78861218bd861632d6cdcac6c8af6a5470

Support should be straight-forward (once the time is right with the appropriate golang release) with the mentioned fallbacks. A likely good practice for adoption would be adding a config option making the feature opt-in first, then given no issues and the community supporting the idea, opt-out could become the default in the future.

Security considerations: Eventually it might be a good idea to support the concept of one host connecting with multiple addresses at the same time. Given the typical usage of Syncthing and its security model, it already seems to be largely independent of relying on hosts using only specific IP addresses, and by starting with opt-in support, security considerations of likely mostly enterprise users would be revealed eventually. Also consider that having multiple addresses is not unique to MPTCP as Syncthing itself is already aware of the concept to some degree, and there are also plans for multipath support in QUIC just on a different layer:
https://datatracker.ietf.org/doc/draft-ietf-quic-multipath/
https://github.com/quic-go/quic-go/issues/3343

**Comment by calmh** [comment]
I see no real downsides here. Mostly it seems to depend on host configuration whether multiple paths are eventually used or not.

Notably, this does not, today, provide multiple connections between single-homed hosts (as is also requested somewhat frequently).

**Comment by imsodin** [comment]
I don't think w need to do anything to get it, it will become the default behaviour eventually. In that case the only question is if we want to opt-in to using it in it's first iteration in go (1.21) or only once deemed working good enough for everyone (sounds like that's expected for 1.22).

Just for context as it came up: I think the typical request for parallel/inverse-multiplexed tcp conns is for low-quality (high roundtrip times and/or packet loss) connections. Typical examples given are all kind of sync/torrent applications that do that.

On first glance at the "multi-path" concept I got super exited, as I thought it would actully try to discover/make use of multiple paths all along the transmission (e.g. different peering combinations from the ISP at one and and the other). That would likely have fully solved the above use-cases. I guess I got a bit overexited there, probably already multiple addresses at the endpoints as in this proposal is complicated enough :)

**Comment by bt90** [comment]
I really need to look more deeply into MPTPC. So far, it's been the graphene of networking. It's great on paper, but it never seemed to make it out of the lab.

How does the user avoid unfavorable routing like e.g via his laptops 5G connection?

**Comment by matttbe** [comment]
@voidpointertonull: thank you for this good idea!

> I really need to look more deeply into MPTCP. So far, it's been the graphene of networking. It's great on paper, but it never seemed to make it out of the lab.

@bt90 it depends on what you consider a lab but MPTCP is used by [Apple with some of their apps](http://blog.multipath-tcp.org/blog/html/2018/12/15/apple_and_multipath_tcp.html) (Siri, Music, Maps, etc.) on quite a few millions of devices. It is also used by quite a large number of people with "[hybrid access networks](https://www.tessares.net/hybrid-access-solution/)".

OK in both cases, the majority of the customers doesn't even know about that but I think we are no longer in a lab case.

> How does the user avoid unfavorable routing like e.g via his laptops 5G connection?

@bt90 It is a question of configuration. If nothing is configured, likely only one path will be used. If you configure it properly, you can mark this costly path as "backup", ask to use it in priority, or attach to existing connections when the primary connection fail, etc.

> While "Go 1.22 or later" does sound like it might be just a matter of time, I'm not that optimistic.

@voidpointertonull: indeed: to change the default behaviour, we will need feedback. It is then a good idea to have such option to get this feedback.

> administrators would get upset about opt-out MPTCP causing only the initiator IP addresses to be logged.

@voidpointertonull: once the (MPTCP) connection is established, you have the same user behind even if other subflows are created. I don't know if you need to track that but technically, the userspace application can get the list of subflows with IP addresses, etc. if needed.

Note that to me, it looks interesting to at least have MPTCP support enabled for the server side. If a "plain TCP" (without MPTCP) connection request is sent to the server, the "accepted" socket will be in "plain TCP": MPTCP stack is bypassed.
For the client side, MPTCP can also be used. If it is not configured, only one path will be used but that's OK, very small overhead but still prepared to support handovers and network aggregation.

> where the new goods are made with Matthieu Baerts doing incredible work seemingly non-stop.

:laughing: I don't think that's correct :stuck_out_tongue: 

(If you say that by looking at my Github stats, that's because our CI is doing a daily sync with upstream and for technical reasons, the last commit has to be set to my name as committer to allow other CIs to do their validation :) )

But nice summary for the rest!

> adding extra links shouldn't degrade performance.

(we are still working on corner cases but that's the idea)

**Comment by calmh** [comment]
> I don't think we need to do anything to get it, it will become the default behaviour eventually. In that case the only question is if we want to opt-in to using it in it's first iteration in go (1.21) or only once deemed working good enough for everyone (sounds like that's expected for 1.22).

I think we do want a knob for this, but the default value should probably match Go's. That means something that can be enabled by early adopters, and later turned off by ... whoever needs to do that.

**Comment by matttbe** [comment]
> > once the (MPTCP) connection is established, you have the same user behind even if other subflows are created. I don't know if you need to track that but technically, the userspace application can get the list of subflows with IP addresses, etc. if needed.
> 
> The program can get all addresses, but by just opting into MPTCP usage it would miss out on everything else but the initiating address, and that's definitely undesired in some cases which is why I don't see opt-out becoming the default for libraries at least for a long time. Surely in some cases I can actually see the need, but also keep enterprise logic in mind where measures are often taken for compliance even if they degrade user experience and in some cases even security (password expiry, SMS 2FA, you get the idea). For example brushing logging needs aside, a platform tying an IP address to a session token would end up regularly invalidating sessions due to the initiator IP address regularly changing if MPTCP would be just kind of forced on it without further support in the program.

It makes sense to have an option to opt-out just in case.

Except if there is an explicit monitoring of the MPTCP "attributes", the IP of the connection doesn't change. Maybe a firewall could block new subflows but I'm not sure to see how MPTCP subflows would be blocked if "plain TCP" connections are allowed.

But yes, there are probably corner cases :)

> > laughing I don't think that's correct stuck_out_tongue
> 
> Hey, it definitely looks that way. I know that there are others from Tessares, but wherever I see MPTCP goods, I mostly see your name, and that includes a ton of support, not just commit formalities. My mind is set, you are making MPTCP adoption possible, and the company is better off keeping your name plastered as a selling point at this point. stuck_out_tongue

Thank you for these kind words! I should forward that to my employer :-D

> > (we are still working on corner cases but that's the idea)
> 
> Surely another reason why it's going to stay opt-in for a while, and it does seem like it's mostly suitable for increasing reliability first and foremost with bandwidth aggregation coming second, still being a bit rough at the edges. Syncthing would mostly benefit from the bandwidth aggregation, but then generally it's also an excellent collection of networking best practices so I'm quite sure that it will help to explore the other potential benefits like the earlier example of figuring whether IPv4 or IPv6 performs better in a given situation.

Indeed. Aggregation works well but for the moment, a very poor primary link can impact the overall performances. We know what kind of optimisation is needed but we need to dedicate time to that.

> Off topic, but I wish there was a VPN solution at least remotely comparable to this project. With performance and user experience varying so heavily based on what's picked from the currently quite fragmented landscape, I occasionally seriously really contemplated dishing out VPN configs through Syncthing in some cases because so far that has been the only really reliable connection maker between hosts, especially when mobility is involved.

(An MPTCP (transparent) proxy might help maybe here)

> > I think we do want a knob for this, but the default value should probably match Go's. That means something that can be enabled by early adopters, and later turned off by ... whoever needs to do that.
> 
> Sounds reasonable, there's really nothing else required as a starter, just letting users easily opt-in. It's not just a significantly better user experience than forcing MPTCP usage with external tools, but it's also more important to have here as the usual LD_PRELOAD approach of forcing doesn't really work with golang, so the bar is higher than elsewhere.

Yes, it would be nice to have an option to force to opt-in (and opt-out) but by default, not doing anything special.

> Eventually it would be great to have the extras like all subflow addresses being shown instead of just the initiator which may be even an address which disappears over the lifetime of the MPTCP connection, but that's really for later.

I guess the Golang "net" package should provide such info at some points.

**Comment by TalalMash** [comment]
I wish to see TCP Options (header) not filtered by many ISPs or sometimes mangled in my case, during travels I noticed that mobile data was mostly unfiltered, while landline did not pass the options field in the majority of time (CGNAT), haven't check IPv6 though.

**Comment by TalalMash** [comment]
> > I wish to see TCP Options (header) not filtered by many ISPs or sometimes mangled in my case, during travels I noticed that mobile data was mostly unfiltered, while landline did not pass the options field in the majority of time (CGNAT), haven't check IPv6 though.
> 
> While I have no direct insight, I believe that filtering that can cause such issues is just becoming more and more common both with CGNAT getting more common, and censorship getting more aggressive as DNS based blocking isn't considered good enough anymore in many places.
> 
> Aside from crappy configurations, even net neutrality likely never truly existed in most countries to begin with. Managed hosts in various countries, and aside from obvious issues with filtering and/or bad routing, I've looked into apt-get not working a bit too many times just to find the received package list actually being a silly localized Vodafone page warning about the data cap just because there's one silly side still considering bare HTTP to be good enough, and another side considering the lack of encryption an open invitation to injection. Anything susceptible to MitM is highly likely to have issues somewhere, there's really no good solution for that without preventing feature downgrade.

To add, TCP options can still be discarded from misconfigured (old) firewall or (old) L3 ASIC/offloading switches.
I've had few dedicated ISP lines and cloud services filtering packets for some reason w/o NAT, e.g Oracle VPC.

**Comment by marbens-arch** [comment]
In Go 1.24, MPTCP is now [enabled by default](https://go.dev/doc/go1.24#netpkgnet).

**Comment by matttbe** [comment]
> In Go 1.24, MPTCP is now [enabled by default](https://go.dev/doc/go1.24#netpkgnet).

That's good news! But only on the listener (server) side.

It would be great to add an option to easily enable it on the dialer (client) side, by simply calling [`Dialer.SetMultipathTCP`](https://pkg.go.dev/net#Dialer.SetMultipathTCP) (available since Go v1.21) :)

(A workaround is to launch syncthing with [`GODEBUG=multipathtcp=1`](https://go.dev/doc/godebug) env var I suppose)

**Comment by obonaventure** [comment]
Today, one of the key benefit of Multipath TCP is on dual-stack hosts that have IPv4 and IPv6. Multipath TCP enables them to automatically select the best performing address family and switch seamlessly from one to another if needed. Clients should use MPTCP by default to benefit from this features
